### PR TITLE
Backport 2.x: PSA storage format: test lifetimes

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
* Cover lifetimes in the PSA storage format tests.
* Reject attempts to create or destroy a read-only key.
* Add a few more test cases around lifetime values.

This is the 2.x version. I'll forward-port to 3.0 once the PR is approved or close.

### CODE REVIEW

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Almost good to me.

File: ChangeLog.d/psa-read-only-keys.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Bugfix
+   * The PSA API no longer allows the creation or destruction of keys with a
+     read-only lifetime. The persistence level PSA_KEY_PERSISTENCE_READ_ONLY
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply ChangeLog.d/psa-read-only-keys.txt:3]: Why? “Key with a read-only lifetime” and “key with a read-only persistence level” are synonymous, but “read-only lifetime” is easier to understand.

[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply ChangeLog.d/psa-read-only-keys.txt:3]: Because in the context of PSA crypto, "read-only" is a persistence level not a lifetime. A lifetime (in the context of PSA crypto) is the combination of a persistence level and a location. I understand that more generally "read-only lifetime" is completely fine but here I find "read-only persistence level" more precise than "read-only lifetime".

File: library/psa_crypto.c [hunk 2]
```diff
@@ -1206,12 +1205,10 @@ psa_status_t psa_destroy_key( mbedtls_svc_key_id_t key )
     }
 #endif /* MBEDTLS_PSA_CRYPTO_SE_C */
 
-#if defined(MBEDTLS_PSA_CRYPTO_SE_C)
 exit:
-#endif /* MBEDTLS_PSA_CRYPTO_SE_C */
     status = psa_wipe_key_slot( slot );
     /* Prioritize CORRUPTION_DETECTED from wiping over a storage error */
-    if( overall_status == PSA_SUCCESS )
+    if( status != PSA_SUCCESS )
```
Comment (ronald-cron-arm) [inline library/psa_crypto.c:1211]: Why did you change that? The idea was to give a previous error code the priority over the `psa_wipe_key_slot` error code.

Comment (gilles-peskine-arm) [reply library/psa_crypto.c:1211]: The comment did not agree with the code. If there was both a storage or permission error (noted in `overall_status`) and a corruption error (noted in `status`), the comment said to return `status` but the code returned `overall_status`.

File: ChangeLog.d/psa-read-only-keys.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply ChangeLog.d/psa-read-only-keys.txt:3]: A key is read-only if its lifetime is read-only. A lifetime is read-only if its persistence level is `PSA_KEY_PERSISTENCE_READ_ONLY`. “Read-only lifetime” is synonymous with “read-only persistence level”, but removing the word “lifetime” from this sentence makes it harder to understand.

[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply ChangeLog.d/psa-read-only-keys.txt:3]: Nevermind, closing this conversation.

File: library/psa_crypto.c (hunk 2 above)
Comment (ronald-cron-arm) [reply library/psa_crypto.c:1211]: ah yes indeed.

**Review by ronald-cron-arm — APPROVED** [review]
Thanks for the changes and explanations. LGTM.

**Review by tom-daubney-arm — APPROVED** [review]
Thanks, this all LGTM. No issues found.
